lib/ecto/pools/ownership.ex

```diff
@@ -0,0 +1,206 @@
+      end
+    end
+
+    def checkout(pool, timeout) do
+      {worker, mod_conn, mode} =
+        GenServer.call(pool, :get_checkout, timeout)
+        |> maybe_raise
+      # TODO: queue_time
+      {:ok, worker, mod_conn, mode, 0}
+    end
+
+    def checkin(_pool, _worker, _timeout) do
+      :ok
+    end
+
+    def break(pool, worker, timeout) do
+      GenServer.call(pool, {:break, worker, timeout}, timeout)
+      |> maybe_raise
+    end
+
+    def checkout_transaction(pool, timeout) do
+      {fun, {worker, mod_conn, mode}} =
+        GenServer.call(pool, {:open_transaction, timeout}, timeout)
+        |> maybe_raise
+
+      fun.()
+      # TODO: queue_time
+      {:ok, worker, mod_conn, mode, 0}
+    end
+
+    def open_transaction(_pool, _worker, _timeout) do
+      raise "#{inspect __MODULE__}.open_transaction/3 should never be called"
+    end
+
+    def close_transaction(pool, worker, timeout) do
+      fun =
+        GenServer.call(pool, {:close_transaction, worker, timeout}, timeout)
+        |> maybe_raise
+
+      fun.()
+    end
+
+    defp maybe_raise({:error, :not_checked_out}),
+      do: raise(ArgumentError, "process doesn't own a worker")
+    defp maybe_raise(:ok),
+      do: :ok
+    defp maybe_raise({:ok, value}),
+      do: value
+
+
+    def init({module, pool}) do
+      {:ok, %{module: module,
+              pool: pool,
+              owners: %{}}}
+    end
+
+    def handle_call({:ownership_checkout, strategy, timeout}, {pid, _ref}, s) do
+      case Map.fetch(s.owners, pid) do
+        {:ok, _} ->
+          {:reply, {:error, :already_checked_out}, s}
+        :error ->
+          checkout = retrieve_worker(timeout, s)
+          {worker, {module, conn}, :default, _queue_time} = checkout
+          mode = if strategy,
+                   do: strategy.ownership_checkout(module, conn),
+                 else: :default
+
+
+          checkout = {worker, {module, conn}, mode}
+          s = monitor_owner(pid, checkout, strategy, s)
+          {:reply, :ok, s}
+      end
+    end
+
+    def handle_call({:ownership_checkin, timeout}, {pid, _ref}, s) do
+      case Map.fetch(s.owners, pid) do
+        {:ok, {_ref, {worker, {module, conn}, _mode}, strategy}} ->
+          if strategy, do: strategy.ownership_checkin(module, conn)
+          s.module.checkin(s.pool, worker, timeout)
+          {:reply, :ok, s}
+        :error ->
+          {:reply, {:error, :not_checked_out}, s}
+      end
+    end
+
+    def handle_call(:get_checkout, {pid, _ref}, s) do
+      maybe_get_worker(pid, s, fn {_ref, checkout, _strategy} ->
+        {:reply, {:ok, checkout}, s}
+      end)
+    end
+
+    def handle_call({:break, worker, timeout}, {pid, _ref}, s) do
+      maybe_get_worker(pid, s, fn {_ref, {my_worker, _mod_conn, _mode}, _strategy} ->
+        ^my_worker = worker
+        s.module.break(s.pool, worker, timeout)
+        # Should we do same as :DOWN here? I am thinking no
```

Comment:
No, we should not checkin. Or the process will die or it will eventually checkin itself. This looks correct afaik. :)